### PR TITLE
Add validation layer for Derefer

### PR DESCRIPTION
**Pull request by ouz-a**
_Follow up work to #96549 #96116 #95857 #95649_

This adds validation for Derefer making sure it is always the first projection.

r? rust-lang/mir-opt

### CODE REVIEW

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/deref_separator.rs:91]: What's the difference between this one and the main one? Can we deduplicate their logic if the differences are small?

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [inline compiler/rustc_mir_transform/src/deref_separator.rs:25]: These checks can save a lot of unnecessary iterations.

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [inline compiler/rustc_mir_transform/src/deref_separator.rs:42]: Although it's small, check here is lighter compared to other function.

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [reply compiler/rustc_mir_transform/src/deref_separator.rs:91]: I added two comments explaining it. I think running `deref_finder` two times is a bit expensive, the new function is a bit less expensive compared to the old one.

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/deref_separator.rs:25]: I think the full version should just have that, too

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/deref_separator.rs:91]: Please just run the full version. We'll benchmark it and if there's a regression we can look into it

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [reply compiler/rustc_mir_transform/src/deref_separator.rs:25]: That makes a lot more sense 😅

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [inline compiler/rustc_mir_transform/src/deref_separator.rs:23]: I think this should remove a lot of unnecessary computation

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/deref_separator.rs:25]:
```suggestion
            && place.projection[1..].contains(&ProjectionElem::Deref)
```

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/deref_separator.rs:33]: This can just iterate over the projection slice instead of iter_projections

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (ouz-a) [inline compiler/rustc_mir_transform/src/deref_separator.rs:33]: I am not sure if this change is going to affect anything at all

File: compiler/rustc_mir_transform/src/deref_separator.rs
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/deref_separator.rs:34]: Instead of indexing here use the second for loop value

**Comment by oli-obk** [comment]
re-rolling as I wrote the change adding a new mir phase (cc @JakobDegen)

r? rust-lang/mir-opt

**Comment by oli-obk** [comment]
@bors try

We should crater this as it adds a new rule to our final MIR output that needs to be upheld or we ICE

File: compiler/rustc_const_eval/src/transform/validate.rs
Comment (JakobDegen) [inline compiler/rustc_const_eval/src/transform/validate.rs:313]: Is there any reason not to use `self.fail` here?

**Comment by oli-obk** [comment]
@craterbot run mode=build-only

need to run in build mode so that we actually run all mir optimizations

**Comment by Mark-Simulacrum** [comment]
@oli-obk just to confirm - debug build is enough to run MIR opts? I don't think crater builds with the release profile today by default.

**Comment by oli-obk** [comment]
> debug build is enough to run MIR opts? I don't think crater builds with the release profile today by default.

yes, it's not important to run any mir opts, it's just important to produce MIR for LLVM

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after resolving other comments

**Comment by JakobDegen** [comment]
Dest prop is affected by this actually, the version in my PR handles it correctly though

File: compiler/rustc_const_eval/src/transform/validate.rs
Comment (ouz-a) [reply compiler/rustc_const_eval/src/transform/validate.rs:313]: It gave horrendous wall of text as error I thought assert's error looked cleaner and concise.

Comment (JakobDegen) [reply compiler/rustc_const_eval/src/transform/validate.rs:313]: Hmm, honestly, I'm not so concerned by the wall of text. This is an ICE after all, it doesn't really have to have good UX, just needs to have all the necessary information. Especially the part about the location being included in the ICE seems really useful.

**Comment by oli-obk** [comment]
crater is clean. So I guess what's left is adding the inliner test and fixing it by running derefer at the end of the inliner

**Comment by ouz-a** [comment]
> There are a few transformation that replace all occurrences of a local with a place. Do they need to be updated if the replacement place might contain any projections, or is it an issue only if projections contain a dereference?
> 
>     * State transform - it seems it is already updated. Ok.
> 
>     * NRVO - replaces only complete locals. Ok.
> 
>     * Inliner - replaces locals with arbitrary places. Fails the new validation. Disabled by default, but a pull requests enabling it waits on bors.
> 
>     * Destination propagation - a replacement can contain projections, but not deref projection. Disabled by default.
> 
> 
> MIR inliner test case
> 
> ```
> $ cat a.rs 
> #![feature(box_syntax)]
> #[inline]
> fn f() -> Box<i32> { box 0 }
> fn main() { box f(); }
> $ rustc +3930b75493eb88cd5431103dbdbc47eb760e9d9b a.rs -Zinline-mir
> thread 'rustc' panicked at '(*(*_7))', compiler/rustc_const_eval/src/transform/validate.rs:314:13
> ```

Thanks, I added this as test and fixed it by running derefer at the end of inline

File: compiler/rustc_const_eval/src/transform/validate.rs
Comment (ouz-a) [reply compiler/rustc_const_eval/src/transform/validate.rs]: I changed it to `self.fail` after trying it out it is indeed more useful than basic assert

**Review by davidtwco — APPROVED** [review]
I'm not sure if this is waiting for my review, but if it is, r=me once CI is passing.

**Comment by GuillaumeGomez** [comment]
Seems like it failed in https://github.com/rust-lang/rust/pull/97496.

@bors r-

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
Please rebase and squash once everything works

**Comment by tmiasko** [comment]
The tests fail on wasm32-unknown-unknown, because this target defaults to `-C panic=abort`, and the function calls in MIR don't have unwind edges:

```diff
-           (*_5) = f() -> [return: bb2, unwind: bb5]; // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
+           (*_5) = f() -> bb2;
```

We have been generally either ignoring test on this target:
 `// ignore-wasm32 compiled with panic=abort by default`
or compiling with panic=abort:
 `// compile-flags: -C panic=abort`.

**Comment by ouz-a** [comment]
> The tests fail on wasm32-unknown-unknown, because this target defaults to `-C panic=abort`, and the function calls in MIR don't have unwind edges:
> 
> ```diff
> -           (*_5) = f() -> [return: bb2, unwind: bb5]; // scope 0 at $DIR/derefer_inline_test.rs:10:9: 10:12
> +           (*_5) = f() -> bb2;
> ```
> 
> We have been generally either ignoring test on this target: `// ignore-wasm32 compiled with panic=abort by default` or compiling with panic=abort: `// compile-flags: -C panic=abort`.

Thanks, going to add `// compile-flags: -C panic=abort` to the test, I was pretty confused why the test was running fine on my machine and CI but failing on merge...

**Pull request merged**
